File: leetcode/Python/word-ladder-ii.py
class Solution:
    def __init__(self):
        self.paths = []
        self.sol = []
    def transform(self, a, b):
        c = 0
        if len(a)!=len(b):
            return False
        for i in range(len(a)):
            if a[i]!=b[i]:
                c+=1
            if c>1:
                return False
        return True if c==1 else False
    def makePaths(self, wordList):
        self.paths = [[] for i in range(len(wordList))]
        for i in range(len(wordList)):
            for j in range(i+1,len(wordList)):
                if self.transform(wordList[i], wordList[j]):
                    self.paths[i].append(j)
                    self.paths[j].append(i)
    def findPaths(self, beginWord, endWord, wordList):
        dq, visit, sol_paths = collections.deque(), [0]*len(wordList), []
        idx, end_idx, s, e = 0, -1, 0, 0
        for i in range(len(wordList)):
            if self.transform(wordList[i], beginWord):
                dq.append(i)
                visit[i]=1
            if endWord == wordList[i]:
                end_idx=i
        if end_idx == -1:
            return
        while len(dq)>0:
            idx=dq.popleft()
            if visit[end_idx]>0 and visit[idx] > visit[end_idx]:
                break
            for next_idx in self.paths[idx]:
                if visit[next_idx] > 0:
                    continue
                dq.append(next_idx)
                visit[next_idx]=visit[idx]+1
        if visit[end_idx]==0:
            return
        sol_paths.append([end_idx])
        while visit[sol_paths[-1][-1]] > 1:
            e = len(sol_paths)
            for i in range(s,e):
                idx = sol_paths[i][-1]
                for next_idx in self.paths[idx]:
                    if visit[next_idx] +1 == visit[idx]:
                        sol_paths.append(sol_paths[i][:])
                        sol_paths[-1].append(next_idx)
            s = e
        for i in range(s,len(sol_paths)):
            t = [beginWord]
            for j in range(len(sol_paths[i])-1,-1,-1):
                t.append(wordList[sol_paths[i][j]])
            self.sol.append(t)
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        self.makePaths(wordList)
        self.findPaths(beginWord, endWord, wordList)
        return self.sol

File: leetcode/C++/word-ladder-ii.cpp
class Solution {
public:
    vector<vector<int>> paths;
    vector<vector<string>> sol;
    bool transform(string &a, string &b){
        int c=0;
        if(a.length() != b.length()) return false;
        for(int i=0; i<a.length()&&c<2; ++i){
            if(a[i]!=b[i]) ++c;
        }
        return (c==1)? true : false;
    }
    void makePaths(vector<string>& wordList){
        paths=vector<vector<int>>(wordList.size());
        for(int i=0; i<wordList.size(); ++i){
            for(int j=i+1; j<wordList.size(); ++j){
                if(transform(wordList[i],wordList[j])){
                    paths[i].push_back(j);
                    paths[j].push_back(i);
                }
            }
        }
    }
    void findPaths(string &beginWord, string &endWord, vector<string>& wordList){
        queue<int> q;
        vector<int> visit(wordList.size(),0);
        vector<vector<int>> sol_paths;
        int idx=0, endIdx=-1, s=0, e=0;
        for(int i=0; i<wordList.size(); ++i){
            if(transform(wordList[i], beginWord)){
                q.push(i);
                visit[i]=1;
            }
            if(endWord==wordList[i]) endIdx=i;
        }
        if(endIdx==-1) return;
        while(!q.empty()){
            idx=q.front(); q.pop();
            if(visit[endIdx]>0 && visit[endIdx]<visit[idx]) break;
            for(int &next_idx : paths[idx]){
                if(visit[next_idx]) continue;
                q.push(next_idx);
                visit[next_idx]=visit[idx]+1;
            }
        }
        if(visit[endIdx]==0) return;
        sol_paths.push_back({endIdx});
        while(visit[sol_paths.back().back()]>1){
            e=sol_paths.size();
            for(int i=s; i<e; ++i){
                idx = sol_paths[i].back();
                for(int &next_idx : paths[idx]){
                    if(visit[next_idx]+1==visit[idx]){
                        sol_paths.push_back(sol_paths[i]);
                        sol_paths.back().push_back(next_idx);
                    }
                }
            }
            s=e;
        }
        for(int i=s; i<sol_paths.size(); ++i){
            vector<string> t;
            t.push_back(beginWord);
            for(int j=sol_paths[i].size()-1; j>=0; --j){
                t.push_back(wordList[sol_paths[i][j]]);
            }
            sol.push_back(t);
        }
    }
    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        makePaths(wordList);
        findPaths(beginWord, endWord, wordList);
        return sol;
    }
};